README.md
<<--Report System-->>

This is the source code for my Report System. If you want more information come here: https://garrysmods.org/download/60644/report-system
If you want to use the workshop version, use this one: http://steamcommunity.com/sharedfiles/filedetails/?id=1297541036


reportsys/CHANGELOG.txt
1.3
- Made reporting players name more hard to change (thanks to DEADMONSTOR and IceGT for alerting me)

1.2
- Added file logging.
- Fixed ULX Jail Bug.

1.1
- Moved file storage to Server to add more protection.
- Added the option to copy reports.
- Added the configurable option to close the panel once a user reports a player.


reportsys/lua/autorun/client/config.lua
Config = {}
--[[

Report System Configuration
By Livaco

]]--

-- Width percentage, ScrW() is the width of the players screen. e.g: If u wanted the width to be 80 percent of the screen, you would put Config.Width = ScrW() * 0.8
Config.Width = ScrW() * 0.8

-- Same above but with the height.
Config.Height = ScrH() * 0.8

-- Server Name, for the DFrame's Title
Config.ServerName = "Example Server"

-- Blur the background of the DFrame or not.
Config.BlurBackground = true

-- What should the default text be in the reason box
Config.DefaultReason = "Reason"

-- If the panel should close after the submit button is pressed. Recomended to stop spam.
Config.ClosePanelOnSubmit = true

-- What it should say underneath the Submit button, sort of like a licence agreement. Can't be too long or it will go off the screen.
Config.SubmitText = "NOTE: Any unfair or spammed reports will result in punishment."

reportsys/lua/autorun/client/cl_adminpanel.lua
--[[

Report System
By Livaco

Don't edit this file if you do not know what your doing. Only edit the configuration.
]]--

local function AdminPanel()

	if LocalPlayer():IsAdmin() then

		-- TODO

		PANEL = {}
		PANEL.Title = " Administrator Menu - Report System by Livaco"
		PANEL.StartTime = SysTime()

		local frame = vgui.Create("DFrame")
		frame:SetSize(Config.Width, Config.Height)
		frame:SetTitle(Config.ServerName .. PANEL.Title)
		frame:SetDraggable(false)
		frame:ShowCloseButton(false)
		frame:Center()
		frame:MakePopup()

		frame.Paint = function(self, w, h)

			if Config.BlurBackground == true then

				Derma_DrawBackgroundBlur(frame, PANEL.StartTime)

			end

			draw.RoundedBox(0, 0, 0, w, h, Color(50, 50, 50))

		end

		local closebutton = vgui.Create("DButton", frame)
		closebutton:SetPos(ScrW() * 0.789, 0)
		closebutton:SetSize(20, 20)
		closebutton:SetText("X")
		closebutton.DoClick = function()

			frame:Close()

		end
		function closebutton:Paint(w, h)

			if closebutton:IsHovered() then

				draw.RoundedBox(0, 0, 0, w, h, Color(50, 50, 50))

			else

				draw.RoundedBox(0, 0, 0, w, h, Color(10, 10, 10))

			end

		end

		local reportlist = vgui.Create("DListView", frame)
		reportlist:Dock(FILL)
		reportlist:AddColumn("Reports - Right Click to Copy.")

		net.Receive("report_reports", function()

				ReportTable = net.ReadTable()

				for k,v in pairs(ReportTable) do

					reportlist:AddLine(v)

				end

		end)

		function reportlist:OnRowRightClick(id, text)

			chat.AddText(Color(120, 120, 120), "[Report System] ", Color(150, 150, 150), "Copied report ID " .. id .. ".")
			SetClipboardText(text:GetColumnText(1))

		end

	end

end

net.Receive("report_openadminpanel", AdminPanel)

reportsys/lua/autorun/server/sv_cmdhandler.lua
--[[

Report System
By Livaco

Don't edit this file if you do not know what your doing. Only edit the configuration
]]--

util.AddNetworkString("report_openuserpanel")
util.AddNetworkString("report_openadminpanel")
util.AddNetworkString("report_nopermerror")
util.AddNetworkString("report_reporthandler")
util.AddNetworkString("report_reports")

reports = {}

net.Receive("report_reporthandler", function(length, ply)

    ReportedPlayer = net.ReadString()
    ReportReason = net.ReadString()

    table.insert(reports, ply:Nick() .. " reported " .. ReportedPlayer .. " for the reason " .. ReportReason)

    local tablejson = util.TableToJSON(reports, false) 
    if not file.IsDir("livaco/reportsystem", "DATA") then

        file.CreateDir("livaco/reportsystem")

    end

    file.Write("livaco/reportsystem/" .. os.date("%m_%d_%Y %I_%M %p") .. ".txt", os.date() .. "\t" .. tablejson)

end)

hook.Add("PlayerSay", "reportsyscmdhandler", function(sender, text, teambool)

	if text == "!report" or text == "!reportplayer" then
		
		net.Start("report_openuserpanel")
		net.Send(sender)

		return ""
		
	end

	if text == "!reportadmin"  then
		
		if sender:IsAdmin() then

			net.Start("report_openadminpanel")
			net.Send(sender)

			net.Start("report_reports")
			net.WriteTable(reports)
			net.Send(sender)

		else

			net.Start("report_nopermerror")
			net.Send(sender)	

		end

	end
	
end)


reportsys/lua/autorun/client/cl_userpanel.lua
--[[

Report System
By Livaco

Don't edit this file if you do not know what your doing. Only edit the configuration.
]]--

IncludeCS("config.lua")

local function UserPanel()

	--chat.AddText(Color(120, 120, 120) .. "[Report System]" .. Color(255, 0, 0) .. " Opening user menu. If you wish to open the admin menu, use !reportadmin")

	PANEL = {}
	PANEL.Title = " User Menu - Report System by Livaco"
	PANEL.StartTime = SysTime()

	ReportedPlayer = "nil"
	ReportReason = "nil"

	local frame = vgui.Create("DFrame")
	frame:SetSize(Config.Width, Config.Height)
	frame:SetTitle(Config.ServerName .. PANEL.Title)
	frame:SetDraggable(false)
	frame:ShowCloseButton(false)
	frame:Center()
	frame:MakePopup()

	frame.Paint = function(self, w, h)

		if Config.BlurBackground == true then

			Derma_DrawBackgroundBlur(frame, PANEL.StartTime)

		end

		draw.RoundedBox(0, 0, 0, w, h, Color(50, 50, 50))

	end

	local closebutton = vgui.Create("DButton", frame)
	closebutton:SetPos(ScrW() * 0.789, 0)
	closebutton:SetSize(20, 20)
	closebutton:SetText("X")
	closebutton.DoClick = function()

		frame:Close()

	end
	function closebutton:Paint(w, h)

		if closebutton:IsHovered() then

			draw.RoundedBox(0, 0, 0, w, h, Color(50, 50, 50))

		else

			draw.RoundedBox(0, 0, 0, w, h, Color(10, 10, 10))

		end

	end

	local players = vgui.Create("DListView", frame)
	players:SetSize(Config.Width * 0.2, Config.Height * 0.9)
	players:SetPos(Config.Width * 0.02, Config.Height * 0.06)
	players:AddColumn("Select Player")

	for k,v in pairs(player.GetAll()) do

		players:AddLine(v:Nick())

	end

	players.OnRowSelected = function(lst, index, pnl)

		ReportedPlayer = pnl:GetColumnText(1)

	end

	local reasontext = vgui.Create("DLabel", frame)
	reasontext:SetPos(Config.Width * 0.25, Config.Height * 0.057)
	reasontext:SetText("Reason:")

	local reasonbox = vgui.Create("DTextEntry", frame)
	reasonbox:SetPos(Config.Width * 0.25, Config.Height * 0.09)
	reasonbox:SetSize(Config.Width * 0.73, Config.Height * 0.02)
	reasonbox:SetUpdateOnType(true)
	reasonbox:SetMultiline(false)
	reasonbox:SetText(Config.DefaultReason)

	reasonbox.OnChange = function(self)

		ReportReason = self:GetValue()

	end

	local submittext = vgui.Create("DLabel", frame)
	submittext:SetPos(Config.Width * 0.25, Config.Height * 0.2)
	submittext:SetSize(Config.Width * 0.73, Config.Height * 0.02)
	submittext:SetText(Config.SubmitText)

	local submitbutton = vgui.Create("DButton", frame)
	submitbutton:SetText("Submit")
	submitbutton:SetPos(Config.Width * 0.25, Config.Height * 0.92)
	submitbutton:SetSize(Config.Width * 0.73, Config.Height * 0.02)

	-- Note: Start a net message and then send the table to the admin panel for it to display.

	submitbutton.DoClick = function()

		net.Start("report_reporthandler")
		net.WriteString(ReportedPlayer)
		net.WriteString(ReportReason)
		net.SendToServer()

		if Config.ClosePanelOnSubmit == true then

			frame:Close()

		end

	end

end 


net.Receive("report_openuserpanel", UserPanel)

